pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+	configPath string
+
+	// holds the mirror sets used for managing registry mirrors.
+	icspRules   map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy
+	idmsRules   map[types.UID]*configV1.ImageDigestMirrorSet
+	itmsRules   map[types.UID]*configV1.ImageTagMirrorSet
+	ruleRWMutex sync.RWMutex
+
+	// timer is used when performing delayed writes to filesystem.
+	timer       *time.Timer
+	timerMutex  sync.Mutex
+	updateDelay time.Duration
+
+	systemContext *ciTypes.SystemContext
+}
+
+var _ Store = (*FileStore)(nil)
+
+// fileStoreOption is used to provide functional options to the fileStore.
+type fileStoreOption func(*FileStore)
+
+// WithConfigPath sets the path to read/write the consolidated registries config.
+func WithConfigPath(path string) fileStoreOption {
+	return func(s *FileStore) { s.configPath = path }
+}
+
+// WithDelay sets a delay before the updated config is written to disk.
+func WithDelay(delay time.Duration) fileStoreOption {
+	return func(s *FileStore) { s.updateDelay = delay }
+}
+
+// NewFileStore creates a new FileStore.
+func NewFileStore(opts ...fileStoreOption) *FileStore {
+	s := &FileStore{
+		configPath:  defaultRegistriesPath,
+		updateDelay: defaultDelay,
+		icspRules:   make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy),
+		idmsRules:   make(map[types.UID]*configV1.ImageDigestMirrorSet),
+		itmsRules:   make(map[types.UID]*configV1.ImageTagMirrorSet),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.systemContext = &ciTypes.SystemContext{SystemRegistriesConfPath: s.configPath}
+	return s
+}
+
+// Cleanup resets the store which includes in-memory and disk resources.
+func (s *FileStore) Cleanup() {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules = make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy)
+	s.idmsRules = make(map[types.UID]*configV1.ImageDigestMirrorSet)
+	s.itmsRules = make(map[types.UID]*configV1.ImageTagMirrorSet)
+
+	if err := os.Remove(s.configPath); err != nil && !os.IsNotExist(err) {
+		log.Warnf("Failed to cleanup registries config at %q: %v", s.configPath, err)
+	}
+	sysregistriesv2.InvalidateCache()
+}
+
+func (s *FileStore) updateConfig() error {
+	if s.updateDelay == 0 {
+		return s.updateConfigNow()
+	}
+
+	return s.updateConfigDelayed()
+}
+
+func (s *FileStore) updateConfigDelayed() error {
+	// Using a different mutex to protect timer vs. rules to avoid deadlock.
+	s.timerMutex.Lock()
+	defer s.timerMutex.Unlock()
+
+	if s.timer != nil {
+		s.timer.Stop()
+	}
+
+	s.timer = time.AfterFunc(s.updateDelay, func() {
+		if err := s.updateConfigNow(); err != nil {
+			log.Errorf("Failed to update the registry mirror config: %v", err)
+		}
+	})
+
+	return nil
+}
+
+func (s *FileStore) updateConfigNow() error {
+	icspRules, idmsRules, itmsRules := s.getAllMirrorSets()
+
+	// Populate config.
+	config := new(sysregistriesv2.V2RegistriesConf)
+	err := registries.EditRegistriesConfig(config, nil, nil, icspRules, idmsRules, itmsRules)
```

Comment:
are we sure those two args should always be `nil`?

Reply:
For the CRs we're currently listening to, yes will always be nil.  The values for those fields can come from multiple places, one such place is the `Image` [CR](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/image-config-openshift-io-v1.html#spec-registrysources).

We're currently not listening to that CR and based on the cases which led to this feature was no mention of using it.  If we need to add that in the future we can.